File: src/puppy_core/puppy_core/dock_detector.py
"""Dock感知闭环 (v7.9)

检测充电桩并执行自动回充流程。
使用LiDAR和视觉传感器融合检测dock位置。

状态机:
  SEARCHING -> DETECTED -> APPROACHING -> ALIGNED -> DOCKED
  失败时: -> RECOVERING -> SEARCHING
"""
import math
import numpy as np
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist, PoseStamped
from sensor_msgs.msg import LaserScan, Image
from std_msgs.msg import String, Bool
from std_srvs.srv import Trigger
from enum import IntEnum
from dataclasses import dataclass
from typing import Optional, Tuple


class DockState(IntEnum):
    """回充状态"""
    IDLE = 0        # 空闲
    SEARCHING = 1   # 搜索dock
    DETECTED = 2     # 检测到dock
    APPROACHING = 3  # 接近中
    ALIGNED = 4      # 已对齐
    DOCKED = 5       # 已对接
    RECOVERING = 6   # 恢复中
    FAILED = 7       # 失败


@dataclass
class DockDetection:
    """Dock检测结果"""
    detected: bool = False
    distance: float = 0.0    # 到dock的距离
    bearing: float = 0.0    # 方位角(弧度)
    confidence: float = 0.0
    method: str = ""        # 'lidar', 'vision', 'fusion'


class DockDetectorNode(Node):
    """Dock感知节点 (v7.9)

    检测充电桩并执行自动回充。
    """

    def __init__(self):
        super().__init__('dock_detector')

        # 参数
        self.declare_parameter('dock_x', 0.0)
        self.declare_parameter('dock_y', -3.5)
        self.declare_parameter('dock_yaw', 0.0)
        self.declare_parameter('approach_speed', 0.1)
        self.declare_parameter('align_threshold', 0.1)  # 对齐阈值(弧度)
        self.declare_parameter('dock_distance_threshold', 0.3)  # 对接距离
        self.declare_parameter('search_timeout', 30.0)  # 搜索超时(秒)

        # dock目标位置
        self.dock_x = self.get_parameter('dock_x').value
        self.dock_y = self.get_parameter('dock_y').value
        self.dock_yaw = self.get_parameter('dock_yaw').value

        # 状态
        self.state = DockState.IDLE
        self._detection = DockDetection()
        self._state_start_time = self.get_clock().now()

        # ROS2接口
        self.create_subscription(LaserScan, '/scan', self._on_scan, 10)
        self.cmd_vel_pub = self.create_publisher(Twist, '/cmd_vel', 10)
        self.state_pub = self.create_publisher(String, '/dock/state', 10)
        self.create_service(Trigger, '/dock/start_docking', self._start_docking)
        self.create_service(Trigger, '/dock/cancel', self._cancel_docking)

        # 控制循环
        self.create_timer(0.1, self._control_loop)
        self.get_logger().info(
            f'Dock感知节点已启动 (v7.9), dock位置: ({self.dock_x:.1f}, {self.dock_y:.1f})')

    def _on_scan(self, msg: LaserScan):
        """处理LiDAR扫描，检测dock特征"""
        # dock特征：特定距离和角度范围内的反射点
        angles = np.arange(len(msg.ranges)) * msg.angle_increment + msg.angle_min
        distances = np.array(msg.ranges)

        # 简化检测：寻找前方近距离的特征点
        front_mask = (np.abs(angles) < 0.5) & (distances < 1.0) & (distances > 0.1)
        if np.any(front_mask):
            front_dists = distances[front_mask]
            front_angles = angles[front_mask]
            # 检测dock的V形特征（两侧有近距离点，中间有远距离点）
            min_dist = np.min(front_dists)
            min_idx = np.argmin(front_dists)
            self._detection = DockDetection(
                detected=True,
                distance=float(min_dist),
                bearing=float(front_angles[min_idx]),
                confidence=0.7,
                method='lidar',
            )
        else:
            self._detection.detected = False

    def _start_docking(self, request, response):
        """启动自动回充"""
        self.state = DockState.SEARCHING
        self._state_start_time = self.get_clock().now()
        response.success = True
        response.message = '回充已启动'
        self.get_logger().info('自动回充已启动')
        return response

    def _cancel_docking(self, request, response):
        """取消回充"""
        self.state = DockState.IDLE
        cmd = Twist()
        self.cmd_vel_pub.publish(cmd)
        response.success = True
        response.message = '回充已取消'
        return response

    def _control_loop(self):
        """主控制循环"""
        if self.state == DockState.IDLE:
            return

        # 发布状态
        state_msg = String()
        state_msg.data = self.state.name
        self.state_pub.publish(state_msg)

        elapsed = (self.get_clock().now() - self._state_start_time).nanoseconds / 1e9

        if self.state == DockState.SEARCHING:
            self._handle_searching(elapsed)
        elif self.state == DockState.DETECTED:
            self._handle_detected()
        elif self.state == DockState.APPROACHING:
            self._handle_approaching()
        elif self.state == DockState.ALIGNED:
            self._handle_aligned()
        elif self.state == DockState.RECOVERING:
            self._handle_recovering(elapsed)

    def _handle_searching(self, elapsed):
        """搜索状态处理"""
        if self._detection.detected:
            self.state = DockState.DETECTED
            self._state_start_time = self.get_clock().now()
            self.get_logger().info(
                f'检测到dock: 距离={self._detection.distance:.2f}m '
                f'方位={math.degrees(self._detection.bearing):.1f}°')
        elif elapsed > self.get_parameter('search_timeout').value:
            self.state = DockState.FAILED
            self.get_logger().error('搜索dock超时')

    def _handle_detected(self):
        """检测到dock状态处理"""
        if not self._detection.detected:
            self.state = DockState.SEARCHING
            return
        self.state = DockState.APPROACHING
        self._state_start_time = self.get_clock().now()

    def _handle_approaching(self):
        """接近dock状态处理"""
        if not self._detection.detected:
            self.state = DockState.SEARCHING
            return

        cmd = Twist()
        align_thresh = self.get_parameter('align_threshold').value
        dock_dist_thresh = self.get_parameter('dock_distance_threshold').value

        # 旋转对准dock
        if abs(self._detection.bearing) > align_thresh:
            cmd.angular.z = 0.5 * np.sign(self._detection.bearing)
        # 前进接近
        elif self._detection.distance > dock_dist_thresh:
            cmd.linear.x = self.get_parameter('approach_speed').value
        else:
            # 到达对接距离
            self.state = DockState.ALIGNED
            self.get_logger().info('已对齐dock，准备对接')

        self.cmd_vel_pub.publish(cmd)

    def _handle_aligned(self):
        """已对齐状态处理"""
        # 缓慢前进完成对接
        if self._detection.distance > 0.1:
            cmd = Twist()
            cmd.linear.x = 0.05
            self.cmd_vel_pub.publish(cmd)
        else:
            self.state = DockState.DOCKED
            cmd = Twist()
            self.cmd_vel_pub.publish(cmd)
            self.get_logger().info('对接完成！')

    def _handle_recovering(self, elapsed):
        """恢复状态处理"""
        if elapsed > 5.0:
            self.state = DockState.SEARCHING
            self._state_start_time = self.get_clock().now()


def main(args=None):
    rclpy.init(args=args)
    node = DockDetectorNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: cpp_src/puppy_core/dock_detector.cpp
// Dock感知闭环 实现 (v7.9)
//
// 检测充电桩并执行自动回充流程
#include "puppy_core/dock_detector.h"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <vector>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

namespace puppy_core {

DockDetectorNode::DockDetectorNode() : rclcpp::Node("dock_detector") {
  // 参数声明
  this->declare_parameter("dock_x", 0.0);
  this->declare_parameter("dock_y", -3.5);
  this->declare_parameter("dock_yaw", 0.0);
  this->declare_parameter("approach_speed", 0.1);
  this->declare_parameter("align_threshold", 0.1);         // 对齐阈值(弧度)
  this->declare_parameter("dock_distance_threshold", 0.3); // 对接距离
  this->declare_parameter("search_timeout", 30.0);         // 搜索超时(秒)

  // dock目标位置
  dock_x_ = this->get_parameter("dock_x").as_double();
  dock_y_ = this->get_parameter("dock_y").as_double();
  dock_yaw_ = this->get_parameter("dock_yaw").as_double();
  approach_speed_ = this->get_parameter("approach_speed").as_double();
  align_threshold_ = this->get_parameter("align_threshold").as_double();
  dock_distance_threshold_ =
      this->get_parameter("dock_distance_threshold").as_double();
  search_timeout_ = this->get_parameter("search_timeout").as_double();

  // 状态初始化
  state_ = DockState::IDLE;
  detection_ = DockDetection();
  state_start_time_ = this->now();

  // ROS2接口
  scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
      "/scan", 10,
      std::bind(&DockDetectorNode::onScan, this, std::placeholders::_1));
  cmd_vel_pub_ =
      this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);
  state_pub_ =
      this->create_publisher<std_msgs::msg::String>("/dock/state", 10);
  start_srv_ = this->create_service<std_srvs::srv::Trigger>(
      "/dock/start_docking",
      std::bind(&DockDetectorNode::startDocking, this, std::placeholders::_1,
                std::placeholders::_2));
  cancel_srv_ = this->create_service<std_srvs::srv::Trigger>(
      "/dock/cancel",
      std::bind(&DockDetectorNode::cancelDocking, this, std::placeholders::_1,
                std::placeholders::_2));

  // 控制循环（10Hz）
  timer_ = this->create_wall_timer(
      std::chrono::milliseconds(100),
      std::bind(&DockDetectorNode::controlLoop, this));

  RCLCPP_INFO(this->get_logger(),
              "Dock感知节点已启动 (v7.9), dock位置: (%.1f, %.1f)",
              dock_x_, dock_y_);
}

void DockDetectorNode::onScan(
    const sensor_msgs::msg::LaserScan::SharedPtr msg) {
  // 处理LiDAR扫描，检测dock特征
  // dock特征：特定距离和角度范围内的反射点

  // 等价 Python:
  //   angles = np.arange(len(msg.ranges)) * msg.angle_increment + msg.angle_min
  //   distances = np.array(msg.ranges)
  //   front_mask = (np.abs(angles) < 0.5) & (distances < 1.0) & (distances > 0.1)
  //   if np.any(front_mask): ...

  const size_t n = msg->ranges.size();
  if (n == 0) {
    detection_.detected = false;
    return;
  }

  // 简化检测：寻找前方近距离的特征点
  // 收集满足 front_mask 条件的点
  std::vector<double> front_dists;
  std::vector<double> front_angles;
  front_dists.reserve(n);
  front_angles.reserve(n);
  for (size_t i = 0; i < n; ++i) {
    double angle = static_cast<double>(i) * msg->angle_increment + msg->angle_min;
    double dist = msg->ranges[i];
    if (std::abs(angle) < 0.5 && dist < 1.0 && dist > 0.1) {
      front_dists.push_back(dist);
      front_angles.push_back(angle);
    }
  }

  if (!front_dists.empty()) {
    // 检测dock的V形特征（两侧有近距离点，中间有远距离点）
    // 等价 Python: min_dist = np.min(front_dists); min_idx = np.argmin(front_dists)
    auto min_it = std::min_element(front_dists.begin(), front_dists.end());
    size_t min_idx = static_cast<size_t>(std::distance(front_dists.begin(), min_it));
    double min_dist = *min_it;

    detection_.detected = true;
    detection_.distance = min_dist;
    detection_.bearing = front_angles[min_idx];
    detection_.confidence = 0.7;
    detection_.method = "lidar";
  } else {
    detection_.detected = false;
  }
}

void DockDetectorNode::startDocking(
    const std_srvs::srv::Trigger::Request::SharedPtr /*request*/,
    std_srvs::srv::Trigger::Response::SharedPtr response) {
  // 启动自动回充
  state_ = DockState::SEARCHING;
  state_start_time_ = this->now();
  response->success = true;
  response->message = "回充已启动";
  RCLCPP_INFO(this->get_logger(), "自动回充已启动");
}

void DockDetectorNode::cancelDocking(
    const std_srvs::srv::Trigger::Request::SharedPtr /*request*/,
    std_srvs::srv::Trigger::Response::SharedPtr response) {
  // 取消回充
  state_ = DockState::IDLE;
  geometry_msgs::msg::Twist cmd;
  cmd_vel_pub_->publish(cmd);
  response->success = true;
  response->message = "回充已取消";
}

void DockDetectorNode::controlLoop() {
  // 主控制循环
  if (state_ == DockState::IDLE) {
    return;
  }

  // 发布状态
  std_msgs::msg::String state_msg;
  state_msg.data = stateName(state_);
  state_pub_->publish(state_msg);

  double elapsed = (this->now() - state_start_time_).seconds();

  switch (state_) {
    case DockState::SEARCHING:
      handleSearching(elapsed);
      break;
    case DockState::DETECTED:
      handleDetected();
      break;
    case DockState::APPROACHING:
      handleApproaching();
      break;
    case DockState::ALIGNED:
      handleAligned();
      break;
    case DockState::RECOVERING:
      handleRecovering(elapsed);
      break;
    default:
      break;
  }
}

void DockDetectorNode::handleSearching(double elapsed) {
  // 搜索状态处理
  if (detection_.detected) {
    state_ = DockState::DETECTED;
    state_start_time_ = this->now();
    RCLCPP_INFO(this->get_logger(),
                "检测到dock: 距离=%.2fm 方位=%.1f°",
                detection_.distance,
                detection_.bearing * 180.0 / M_PI);
  } else if (elapsed > search_timeout_) {
    state_ = DockState::FAILED;
    RCLCPP_ERROR(this->get_logger(), "搜索dock超时");
  }
}

void DockDetectorNode::handleDetected() {
  // 检测到dock状态处理
  if (!detection_.detected) {
    state_ = DockState::SEARCHING;
    return;
  }
  state_ = DockState::APPROACHING;
  state_start_time_ = this->now();
}

void DockDetectorNode::handleApproaching() {
  // 接近dock状态处理
  if (!detection_.detected) {
    state_ = DockState::SEARCHING;
    return;
  }

  geometry_msgs::msg::Twist cmd;
  double align_thresh = align_threshold_;
  double dock_dist_thresh = dock_distance_threshold_;

  // 旋转对准dock
  // 等价 Python: cmd.angular.z = 0.5 * np.sign(self._detection.bearing)
  if (std::abs(detection_.bearing) > align_thresh) {
    double sign = (detection_.bearing > 0.0) ? 1.0
                  : (detection_.bearing < 0.0) ? -1.0
                  : 0.0;
    cmd.angular.z = 0.5 * sign;
  }
  // 前进接近
  else if (detection_.distance > dock_dist_thresh) {
    cmd.linear.x = approach_speed_;
  } else {
    // 到达对接距离
    state_ = DockState::ALIGNED;
    RCLCPP_INFO(this->get_logger(), "已对齐dock，准备对接");
  }

  cmd_vel_pub_->publish(cmd);
}

void DockDetectorNode::handleAligned() {
  // 已对齐状态处理
  // 缓慢前进完成对接
  if (detection_.distance > 0.1) {
    geometry_msgs::msg::Twist cmd;
    cmd.linear.x = 0.05;
    cmd_vel_pub_->publish(cmd);
  } else {
    state_ = DockState::DOCKED;
    geometry_msgs::msg::Twist cmd;
    cmd_vel_pub_->publish(cmd);
    RCLCPP_INFO(this->get_logger(), "对接完成！");
  }
}

void DockDetectorNode::handleRecovering(double elapsed) {
  // 恢复状态处理
  if (elapsed > 5.0) {
    state_ = DockState::SEARCHING;
    state_start_time_ = this->now();
  }
}

std::string DockDetectorNode::stateName(DockState s) {
  // 对应 Python: self.state.name（枚举名）
  switch (s) {
    case DockState::IDLE:        return "IDLE";
    case DockState::SEARCHING:   return "SEARCHING";
    case DockState::DETECTED:    return "DETECTED";
    case DockState::APPROACHING: return "APPROACHING";
    case DockState::ALIGNED:     return "ALIGNED";
    case DockState::DOCKED:      return "DOCKED";
    case DockState::RECOVERING:  return "RECOVERING";
    case DockState::FAILED:      return "FAILED";
  }
  return "UNKNOWN";
}

}  // namespace puppy_core

// ===== 节点入口 =====
int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<puppy_core::DockDetectorNode>());
  rclcpp::shutdown();
  return 0;
}
